Step2 연관 관계 맵핑 리뷰 요청 드립니다.!

lcjltj opened the pull request:
안녕하세요.  Step2 과제 인 연관 관계 맵핑. 리뷰 요청 드립니다 ㅠㅠ.

xxxToOne 의 fetch 전략을 default 값이 Eager로 알고 있어서, 지금 과제에서는 Lazy로 변경 하였습니다. !!

현재는 다대일 단방향 맵핑만 한상태 입니다.

혹시나 잘못된 부분이나, 빼먹은 부분이 있으면 ㅠㅠ 지적 부탁드립니다. JPA 자체가 처음이라 너무  생소해서 적응이 안되네요 ㅠㅠ

lcjltj committed:
refactor(Question): 작성자 필드 User 클래스 연관 관계 설정

lcjltj committed:
docs: 연관 관게 맵핑 요구 사항 정의

lcjltj committed:
refactor(Question): Question writer 연관 관계 맵핑

lcjltj committed:
refactor(Answer): Answer writer, question 연관 관계 맵핑

lcjltj committed:
refactor(DeleteHistory): DeleteHistory deleteBy 연관 관계 맵핑

lcjltj committed:
docs: 연관 관계 맵핑 업데이트

joojimin commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -11,10 +11,12 @@ public class Answer extends BaseTimeEntity {
-    private Long writerId;
-    @Column(name = "question_id")
-    private Long questionId;
+    @ManyToOne(fetch = FetchType.LAZY)
```
연관관계 매핑 👍 👍

joojimin commented on src/main/java/qna/readme.md:
```diff
@@ -0,0 +1,9 @@
+- [x] answer 의 question_id 연관 관계 맵핑
+- [x] answer 의 writer_id 연관 관계 맵핑
+- [x] delete_history deleted_by_id 연관 관계 맵핑
+- [x] question 의 writer_id 연관 관계 맵핑
```
요구사항정리 잘하셧습니다!
다만 일반적으로 README.md 파일은 자바 패키지와 상관없이 최상단 레벨 패키지에 두는걸 관례로 하고있습니다!

lcjltj replied:
README 파일 이동 했습니다!!

joojimin commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -48,10 +48,10 @@ public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteE
         List<DeleteHistory> deleteHistories = new ArrayList<>();
         question.setDeleted(true);
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriterId(), LocalDateTime.now()));
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, loginUser, LocalDateTime.now()));
```
`new DeleteHistory(ContentType.QUESTION, questionId, loginUser, LocalDateTime.now()`
`new DeleteHistory(ContentType.ANSWER, answer.getId(), loginUser, LocalDateTime.now())`
두가지 생성자가 Enum값을 기준으로 구분되는대요! 정적 팩토리 메서드를 이용해 toAnswer() 등으로 네이밍을 주면 잘못된 Enum값이 셋팅될 확률도 줄어들거같습니다!

joojimin replied:
추가적으로 `List<DeleteHistory> deleteHistories` 을 관리하는 일급컬렉션을 만들어 책임을 분리해보면 어떨까요~?

lcjltj replied:
팩토리 메서드 추가 및 일급 컬렉션 적용 헀습니다.!

joojimin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -12,12 +12,14 @@ public class Question extends BaseTimeEntity {
+    private User writer;
     @Column(nullable = false)
     private boolean deleted = false;
 
```
객체 지향적인 설계를 연습하고 있다보니 Answer <-> Question간의 양방향 매핑을 통해 Answer을 로드해보면 어떨까요?

lcjltj replied:
양방향 매핑을 적용하긴 했는데 잘못한 부분이있는지 피드백 부탁드립니다.
JPA를 경험 해보지 않았고 실무에서는 테이블 자체가 이런식으로 분리 되어 있지않아 생각보다 많이 어렵네요.
주말 동안 혼자 프로젝트로 이것 저것 해보고 적용해 보긴 했지만. 막막합니다. ㅠㅠ

joojimin reviewed (changes requested):
안녕하세요~ 찬준님!
연관관계 매핑 요구사항대로 잘해주셧습니다! 
다만 몇가지 빡빡할수도있는 피드백 드렸는데요! 확인해보시고
리뷰 재요청 부탁드립니다!

lcjltj committed:
docs: readme 위치 변경

lcjltj committed:
refactor: Question User 양방향 맵핑

lcjltj committed:
refactor: DeleteHistory 팩토리패턴 적용 및 일급 콜렉션 적용

lcjltj committed:
test: User Question 양방향 연관 관계 테스트 케이스 추가

joojimin commented on src/main/java/qna/domain/Question.java:
```diff
@@ -72,8 +72,14 @@ public User getWriter() {
+        }
+
+        if (!writer.getQuestion().contains(this)) {
+            writer.getQuestion().add(this);
```
List를 꺼내서 add하기보다 add하는 행위를 User 도메인으로 옮겨보면 어떨까요~?

joojimin commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,45 @@
+        this.histories = histories;
+    }
+
+    public void add(final DeleteHistory deleteHistory) {
```
`List<DeleteHistory> histories`를 받아 생성자에서 셋팅을 시켜준다면 add method와 `Collections.unmodifiableList(histories);`를 없애고 처음부터 unmodifiableList로 만들어두면 어떨까요~?

joojimin reviewed (changes requested):
안녕하세요~ 찬준님!
피드백 반영해주시느냐고 고생하셧습니다!!
대부분 반영 잘해주셧는데 다음미션가기전에 몇가지 생각해보고 가면 좋을 이슈들이 있어서 
다시 RC드렸어요~ 확인해보시고 리뷰재요청부탁드려요!

lcjltj committed:
refactor(User):  User containQuestion 메소드 추가
- containQuestion() 질문 포함 유무 테스트 케이스 추가
- update() -> updateNameAndEmail() 메소드 명 변경
    - updateNameAndEmail()
    - equalsNameAndEmail() 테스트 케이스 추가

lcjltj committed:
refactor(User): Question 일급 컬렉션 적용 외 1

- List<Question> -> Questions
    - OnetoMany List 일급 컬렉션 적용을 위해 @Embedded 사용
- UserRepositoryTest 와 UserTest 관심사 분리
    - User 에서 테스트 되야 할 내용 분리

lcjltj committed:
test(Answer): Answer 테스트 케이스 추가

lcjltj committed:
fix: static 유저 사용 시 여러 개의 테스트 실행시 테스트 깨짐 수정

lcjltj committed:
refactor: DeleteHistories default 생성자 제거 외 1

- DeleteHistories
    - default 생성자 제거
    - add() method 제거
    - histories 생성 시 unmodifiableList 초기 생성

joojimin reviewed (approved):
피드백 반영 고생하셧습니다!
다음미션 진행부탁드려요~
